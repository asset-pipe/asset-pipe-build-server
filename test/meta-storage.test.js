'use strict';

const MetaStorage = require('../lib/meta-storage');
const SinkMem = require('@asset-pipe/sink-mem');

test('constructor() - should take a sink', async () => {
    const sink = new SinkMem();
    // eslint-disable-next-line no-new
    new MetaStorage(sink);
});

test('constructor() - should require a sink', async () => {
    expect(() => {
        // eslint-disable-next-line no-new
        new MetaStorage();
    }).toThrowError('MetaStorage needs a sink');
});

test('get() - should reject if missing entry', async () => {
    expect.assertions(1);

    const sink = new SinkMem();
    const storage = new MetaStorage(sink);

    try {
        await storage.get('key');
    } catch (e) {
        expect(e).toEqual(
            new Error('No file could be located with name "/meta/key.json".')
        );
    }
});

test('get() - should error if non-valid json', async () => {
    expect.assertions(1);
    const sink = new SinkMem();
    const storage = new MetaStorage(sink);

    sink.db[storage.getKey('key')] = 'non-valid-json';

    try {
        await storage.get('key');
    } catch (e) {
        expect(e).toEqual(new Error('Failed parsing payload from key "key"'));
    }
});

test('get() - should error if no payload', async () => {
    expect.assertions(1);
    const sink = new SinkMem();
    const storage = new MetaStorage(sink);

    sink.db[storage.getKey('key')] = '';

    try {
        await storage.get('key');
    } catch (e) {
        expect(e).toEqual(
            new Error('No file could be located with name "/meta/key.json".')
        );
    }
});

test('get() - should error if no result and no sink error', async () => {
    expect.assertions(1);
    const sink = {
        async get() {
            return null;
        },
    };
    const storage = new MetaStorage(sink);

    try {
        await storage.get('key');
    } catch (e) {
        expect(e).toEqual(
            new Error(`Failed getting stored payload from key "key"`)
        );
    }
});

test('set() - set and get same value', async () => {
    const sink = new SinkMem();
    const storage = new MetaStorage(sink);

    const random = Math.random() * 1000;

    await storage.set('key', {
        random,
    });

    const result = await storage.get('key');
    expect(result.random).toBe(random);
});

test('has() - set and get same value', async () => {
    const sink = new SinkMem();
    const storage = new MetaStorage(sink);

    await storage.set('key', {});

    const result = await storage.has('key');
    expect(result).toBe(true);
});
