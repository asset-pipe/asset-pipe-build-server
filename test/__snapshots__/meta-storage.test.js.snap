// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all() - error when empty 1`] = `"Missing folder with name \\"/meta/random-type-123\\" or empty result"`;

exports[`all() - return 1 file 1`] = `
Array [
  Object {
    "content": "{\\"foo\\":\\"bar\\"}",
    "fileName": "/meta/text/key.json",
  },
]
`;

exports[`all() - return files from given folder 1`] = `
Array [
  Object {
    "content": "{\\"foo\\":\\"bar\\"}",
    "fileName": "/meta/text/key.json",
  },
  Object {
    "content": "{\\"foo\\":\\"baz\\"}",
    "fileName": "/meta/text/key2.json",
  },
]
`;

exports[`all() - return mulitiple files 1`] = `
Array [
  Object {
    "content": "{\\"foo\\":\\"bar\\"}",
    "fileName": "/meta/text/key.json",
  },
  Object {
    "content": "{\\"foo\\":\\"baz\\"}",
    "fileName": "/meta/text/key2.json",
  },
  Object {
    "content": "{\\"foo\\":\\"bazoooka\\"}",
    "fileName": "/meta/text/key3.json",
  },
]
`;
